File: stack.py
class Node:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next

    def __str__(self):
        return "Node[Data=%s]" % self.data

class Stack:
    def __init__(self, data=[]) -> None:
        self.head = None
        if data:
            for data in data:
                self.push(data)
    def push(self, data):
        temp = Node()
        temp.data = data
        temp.next = self.head
        self.head = temp

    def pop(self):
        if self.head is None:
            raise IndexError
        temp = self.head.data
        self.head = self.head.next
        return temp

    def peek(self):
        if self.head is None:
            raise IndexError
        return self.head.data

    def isEmpty(self):
        return self.head is None

    @staticmethod
    def delimitorIsOk(expression: str) -> bool:
        stack = Stack([])
        for char in expression:
            if char in "([{":
                stack.push(char)
            if char in ")]}":
                try:
                    delimiter = stack.pop()
                except:
                    return False
                if delimiter == "(" and char != ")":
                    return False
                if delimiter == "{" and char != "}":
                    return False
                if delimiter == "[" and char != "]":
                    return False
        return stack.isEmpty()

    @staticmethod
    def htmlIsOk(html: str) -> bool:
        stack = Stack()
        count = 0

        while count < len(html):
            start = html.find('<', count)
            if start == -1:
                break

            end = html.find('>', start + 1)
            if end == -1:
                return False

            tag = html[start + 1:end]

            if not tag.startswith('/'):
                stack.push(tag)
            else:
                if stack.isEmpty() or stack.pop() != tag[1:]:
                    return False

            count = end + 1

        return stack.isEmpty()

stack = Stack([3, 2, 1])

print("Testando pilha:")
print(stack.isEmpty())
print(stack.pop())
print(stack.pop())
print(stack.pop())
print("-------------------------")
print("delimitadorIsOk():")
print(Stack.delimitorIsOk("(1+2)*(5-1)"))
print(Stack.delimitorIsOk("(1+2]-1"))
print(Stack.delimitorIsOk("[((1+2)-(1-4)"))
print("-------------------------")
print("htmlIsOk():")
print(Stack.htmlIsOk("<tag></tag><></>"))
print(Stack.htmlIsOk("<html><body><p>Texto</p></body></html>"))
print(Stack.htmlIsOk("<html><body><p>Texto</body></html>"))
print("-------------------------")

File: stack.go
package main

import (
	"fmt"
	"strings"
)

type Node struct {
	Data string
	Next *Node
}

type Stack struct {
	Head *Node
}

func (s *Stack) push(data string) {
	temp := Node{Data: data, Next: s.Head}
	s.Head = &temp
}

func (s *Stack) pop() (string, error) {
	if s.Head == nil {
		return "0", fmt.Errorf("Pilha vazia")
	}
	data := s.Head.Data
	s.Head = s.Head.Next
	return data, nil
}

func (s *Stack) peek() (string, error) {
	if s.Head == nil {
		return "0", fmt.Errorf("Pilha vazia")
	}
	return s.Head.Data, nil
}

func (s *Stack) is_empty() bool {
	return s.Head == nil
}

func delimiterIsOk(expression string) bool {
	stack := Stack{}
	for _, char := range expression {
		if strings.Contains("({[", string(char)) {
			stack.push(string(char))
		}
		if strings.Contains(")}]", string(char)) {
			data, err := stack.pop()
			if err != nil {
				return false
			}
			if data == "(" && string(char) != ")" {
				return false
			}
			if data == "[" && string(char) != "]" {
				return false
			}
			if data == "{" && string(char) != "}" {
				return false
			}
		}
	}
	return true
}

func htmlIsOk(html string) bool {
	stack := Stack{}
	var count int
	for count <= len(html) {
		start := findInText(html, "<", count)
		end := findInText(html, ">", start+1)

		if start < 0 || end < 0 {
			break
		}

		tag := html[start+1 : end]

		if !strings.HasPrefix(tag, "/") {
			stack.push(tag)
		} else {
			top, err := stack.pop()

			if err != nil {
				return false
			}
			if top != tag[1:] {
				return false
			}
		}
		count = end + 1
	}
	return stack.is_empty()
}
func findInText(text string, target string, start int) int {
	for i := start; i < len(text); i++ {
		if target == string(text[i]) {
			return i
		}
	}
	return -1
}

func main() {
	stack := Stack{}
	stack.push("3")
	stack.push("2")
	stack.push("1")
	fmt.Println(stack.peek())
	fmt.Println(stack.pop())
	fmt.Println(stack.pop())
	fmt.Println(stack.pop())

	fmt.Println()
	fmt.Println(htmlIsOk("<html><body><p>Texto</p></body></html>"))
	fmt.Println(htmlIsOk("<html><body><p>Texto</body></html>"))
}
